Fix and improve OriginValidator

Chukak opened the pull request:
See issue  #929 

Added: 
* methods
    * `self.valid_origin_full(self, origin)` 
    * `self.validate_host_full(self, origin)`
    * `self.is_allowed_origin(self, origin, pattern)`
    * `self.get_origin_ports(self, origin)`
    * `self.check_pattern(self, origin, pattern)`
* attributes
    * `self.full`
    * `self.check_cert`


The ability to check only domains (by default) or to check the entire source is added. You can add both complete and incomplete sources to the list of allowed domains, but for complete verification, at least a schema and domain are required. You can also check the ssl certificate.

Also included are tests and an update for the documentation.

bluetech commented:
By an unfortunate coincidence, I have just written [my own version](https://github.com/bluetech/channels/commits/fix-origin-validator) of this change. Since you already submitted, I'll tuck mine away, though you can take a look and see if there's something helpful there.

Thanks for working on the issue!

andrewgodwin commented on channels/security/websocket.py:
```diff
@@ -10,26 +12,38 @@ class OriginValidator:
     """
 
-    def __init__(self, application, allowed_origins):
+    def __init__(self, application, allowed_origins, full=False, check_cert=False):
```
Can we have a better name for this than `full`?

Chukak replied:
Can you suggest a better name than `` full``, maybe ``origin`` or ``scheme``. Is it necessary to leave the possibility of verifying the certificate or include it in the full check origin.

andrewgodwin commented on channels/security/websocket.py:
```diff
@@ -40,9 +54,88 @@ def valid_origin(self, origin):
+        # Only if not full: https://domain.example.com:8443 -> domain.example.com
+        #   or //domain.example.com - > domain.example.com
+        #   or .example.com -> .example.com
+        self.allowed_origins = [urlparse(pattern).hostname or urlparse("//" + pattern).hostname
```
Can you rewrite this without using `or`? Two in a row makes it harder to skim over.

andrewgodwin commented on channels/security/websocket.py:
```diff
@@ -40,9 +54,88 @@ def valid_origin(self, origin):
+        #   or //domain.example.com - > domain.example.com
+        #   or .example.com -> .example.com
+        self.allowed_origins = [urlparse(pattern).hostname or urlparse("//" + pattern).hostname
+                                or pattern for pattern in self.allowed_origins]
```
The current code indent style is one level and the trailing bracket on the next line - please try and match that!

andrewgodwin commented on channels/security/websocket.py:
```diff
@@ -10,26 +12,38 @@ class OriginValidator:
-        # Check to see if the origin header is valid
-        if self.valid_origin(origin_host):
+        # Check to see if the origin header is valid. Full or default check
+        valid = self.valid_origin(origin_hostname) if not self.full else self.valid_origin_full(result_parse)
```
I'd rather do this check inside a single `valid_origin` function.

andrewgodwin commented on channels/security/websocket.py:
```diff
@@ -40,9 +54,88 @@ def valid_origin(self, origin):
+        # if origin.port doesn`t exists
+        if origin.scheme == "http":
+            # Default ports return for http
+            return 80, 8080
```
Is port 8080 actually specified as a default HTTP port anywhere?

hundredeir replied:
Yes

Chukak commented:
I updated my code. Now, ``OriginValidator`` does not require any options. In ``allowed_origins``, you can pass only domain or origin as scheme://domain[:port] (for exapmple,  ``.domain.com``, ``*``, ``http://.domain.com``.)

andrewgodwin commented on channels/security/websocket.py:
```diff
@@ -21,27 +24,134 @@ def __call__(self, scope):
+        After the scheme there must be a domain. Any domain beginning with a period
+        corresponds to the domain and all its subdomains (for example, ``http://.example.com``
+        ``http://example.com`` and any subdomain).After the domain there must be a port,
+        but it can be omitted. `` * `` matches anything and anything
```
No spaces around the `*` here I think.

andrewgodwin commented on channels/security/websocket.py:
```diff
@@ -21,27 +24,134 @@ def __call__(self, scope):
+        specified list `` allowed_origins``. Any pattern begins with a scheme.
+        After the scheme there must be a domain. Any domain beginning with a period
+        corresponds to the domain and all its subdomains (for example, ``http://.example.com``
+        ``http://example.com`` and any subdomain).After the domain there must be a port,
```
Space after the `.` needed

andrewgodwin commented on channels/security/websocket.py:
```diff
@@ -21,27 +24,134 @@ def __call__(self, scope):
+            for pattern in self.allowed_origins
+        )
+
+    def is_allowed_origin(self, parsed_origin, pattern):
```
I would call this `match_allowed_origin` so it's obviously different from `validate_origin`

andrewgodwin commented on channels/security/websocket.py:
```diff
@@ -21,27 +24,134 @@ def __call__(self, scope):
+        # Compares hostname, scheme, ports of pattern and origin
+        if parsed_pattern.scheme == parsed_origin.scheme and origin_port == pattern_port \
+                and is_same_domain(parsed_origin.hostname, parsed_pattern.hostname):
+            # Check ssl security certificate
```
I do not think we should try to check the certificate in this code at all, and definitely not by default.

Chukak commented:
I updated the code in accordance with your recommendations. Yes, you were right, `` * .example.com`` is a backwards incompatible change, I added tests for this.

andrewgodwin commented on docs/topics/security.rst:
```diff
@@ -37,10 +39,17 @@ valid domains as the second argument::
 
+Note: If you want to resolve any domains, just add ``["*"]`` as second argument.
+
+``OriginValidator`` can accept the ``check_cert`` as fourth argument.
```
`check_cert` is no longer there

andrewgodwin commented on docs/topics/security.rst:
```diff
@@ -25,7 +25,9 @@ ASGI middlewares it contains, ``OriginValidator`` and
 header that is sent with every WebSocket to say where it comes from. Just wrap
 it around your WebSocket application code like this, and pass it a list of
-valid domains as the second argument::
+valid domains as the second argument. You can pass only domain (for example,
```
"only a single domain"

andrewgodwin commented on docs/topics/security.rst:
```diff
@@ -25,7 +25,9 @@ ASGI middlewares it contains, ``OriginValidator`` and
 it around your WebSocket application code like this, and pass it a list of
-valid domains as the second argument::
+valid domains as the second argument. You can pass only domain (for example,
+``.allowed-domain.com``) and full origin, than must be scheme://domain[:port]
```
"and full origin, than must be" -> "or a full origin, in the format" (and make the format code-style formatting)

andrewgodwin commented on docs/topics/security.rst:
```diff
@@ -37,10 +39,17 @@ valid domains as the second argument::
         ),
     })
 
+Note: If you want to resolve any domains, just add ``["*"]`` as second argument.
```
"any domains, just add * as second argument" -> "any domain, then use the origin `*`"

Chukak commented:
I updated the docs for websockets.

andrewgodwin commented:
Thanks so much!
